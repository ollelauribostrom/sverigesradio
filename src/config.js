const endpointURL = 'https://api.sr.se/api/v2/channels?format=json';

export { endpointURL };
