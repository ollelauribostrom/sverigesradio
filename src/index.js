export { listen } from './listen';
export { createChannelProvider } from './channels';
