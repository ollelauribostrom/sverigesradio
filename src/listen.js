import request from 'request';
import Speaker from 'speaker';
import { Decoder, STEREO } from 'lame';

export function listen(streamUrl) {
  const decoder = new Decoder();
  const speaker = new Speaker({
    channels: 2,
    bitDepth: 16,
    sampleRate: 44100,
    mode: STEREO,
  });
  request
    .get(streamUrl)
    .pipe(decoder)
    .pipe(speaker);
}
