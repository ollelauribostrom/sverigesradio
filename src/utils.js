export function formatName(str) {
  return str.split(' ').join('-').toUpperCase();
}

export function parseChannels(channels) {
  return channels.reduce((channelObj, next) => {
    channelObj[formatName(next.name)] = next;
    return channelObj;
  }, {});
}
