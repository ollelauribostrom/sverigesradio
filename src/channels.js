import request from 'request';
import { channelEndpoint } from './config';
import { parseChannels } from './utils';

export function createRequestError(response) {
  const error = new Error();
  error.statusCode = response && response.statusCode;
  return error;
}

export function getChannels() {
  return new Promise((resolve, reject) => {
    request.get(channelEndpoint, (err, response, body) => {
      if (err) {
        reject(createRequestError(response));
      }
      if (response && response.statusCode !== 200) {
        reject(createRequestError(response));
      }
      const { channels } = JSON.parse(body);
      resolve(parseChannels(channels));
    });
  });
}

export async function createChannelProvider() {
  const channels = await getChannels();
  return {
    active: channels.P3,
    select(channelName = '') {
      this.active = channels[channelName.toUpperCase()] || channels.P3;
    },
  };
}
