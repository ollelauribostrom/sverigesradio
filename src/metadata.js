import request from 'request';
import { scheduleEndpoint, songEndpoint } from './config';
import { parseDate } from './utils';

export function getMetadata(url) {
  return new Promise((resolve, reject) => {
    request.get(url, (error, response, body) => {
      if (error) {
        reject();
      }
      if (response && response.statusCode !== 200) {
        reject();
      }
      resolve(JSON.parse(body));
    });
  });
}

export async function updateCurrentProgram(provider) {
  try {
    const url = `${scheduleEndpoint}&channelid=${provider.channelId}`;
    const { channel } = await getMetadata(url);
    const { currentscheduledepisode: program } = channel;
    const duration = parseDate(program.endtimeutc) - Date.now();
    provider.currentProgram = program.title;
    provider.onChangeListeners.forEach(fn => fn());
    setTimeout(() => updateCurrentProgram(provider), duration);
  } catch (err) {
    provider.currentProgram = '';
    setTimeout(() => updateCurrentProgram(provider), 1000 * 60);
  }
}

export async function updateCurrentSong(provider) {
  try {
    const url = `${songEndpoint}&channelid=${provider.channelId}`;
    const { playlist } = await getMetadata(url);
    const { song, nextsong } = playlist;
    if (song) {
      const endtime = parseDate(song.stoptimeutc);
      const now = Date.now();
      if (endtime > now) {
        provider.currentSong = song.description;
        setTimeout(() => updateCurrentSong(provider), endtime - now);
      } else {
        provider.currentSong = '';
        const duration = nextsong ? parseDate(nextsong.starttimeutc) - now : 1000 * 10;
        setTimeout(() => updateCurrentSong(provider), duration);
      }
    } else {
      provider.currentSong = '';
      const duration = parseDate(nextsong.starttimeutc) - Date.now();
      setTimeout(() => updateCurrentSong(provider), duration);
    }
    provider.onChangeListeners.forEach(fn => fn());
  } catch (err) {
    provider.currentSong = '';
    setTimeout(() => updateCurrentSong(provider), 1000 * 60);
  }
}

export function createMetadataProvider(channelId) {
  const provider = {
    channelId,
    currentProgram: '',
    currentSong: '',
    onChangeListeners: [],
    onChange(fn) {
      this.onChangeListeners.push(fn);
    },
    setChannel(newChannelId) {
      this.channelId = newChannelId;
      updateCurrentProgram(this);
      updateCurrentSong(this);
    },
  };
  updateCurrentProgram(provider);
  updateCurrentSong(provider);
  return provider;
}
