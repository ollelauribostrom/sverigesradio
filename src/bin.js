#! /usr/bin/env node

import commander from 'commander';
import clear from 'clear';
import ora from 'ora';
import { listen } from './listen';
import { createChannelProvider } from './channels';
import { version } from '../package.json';

export async function run(argv) {
  try {
    const channels = await createChannelProvider();
    commander
      .version(version)
      .arguments('[channelName]')
      .action(channel => channels.select(channel))
      .parse(argv);

    listen(channels.active.liveaudio.url);
    clear();
    ora(`Sveriges Radio ${channels.active.name}`).start();
  } catch (err) {
    if (err.statusCode) {
      ora(`Sveriges Radio is currently not responding (Status Code: ${err.statusCode})`).fail();
      return;
    }
    ora('Could not connect to Sveriges Radio. Please check your connection.').fail();
  }
}

if (process.env.NODE_ENV !== 'test') {
  run(process.argv);
}
