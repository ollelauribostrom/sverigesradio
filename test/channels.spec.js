import request from 'request';
import { createChannelProvider, getChannels } from '../src/channels';
import mockResponse from './resources/mockResponse';

jest.mock('request', () => ({
  get: jest.fn((url, cb) => cb(null, { statusCode: 200 }, mockResponse)),
}));

describe('Tests for channels.js', () => {
  describe('createChannelProvider', () => {
    it('should create a channel provider', async () => {
      const channelProvider = await createChannelProvider();
      expect(channelProvider.active.name).toEqual('P3');
    });
    it('should update the active channel on select', async () => {
      const channelProvider = await createChannelProvider();
      channelProvider.select('p1');
      expect(channelProvider.active.name).toEqual('P1');
    });
    it('should default the active channel to P3 when passing non existing channel name', async () => {
      const channelProvider = await createChannelProvider();
      channelProvider.select('p1');
      channelProvider.select('Non existing channel');
      expect(channelProvider.active.name).toEqual('P3');
    });
    it('should default the active channel to P3 when not passing any channel argument', async () => {
      const channelProvider = await createChannelProvider();
      channelProvider.select('p1');
      channelProvider.select();
      expect(channelProvider.active.name).toEqual('P3');
    });
  });
  describe('getChannels', () => {
    it('should return a channelObject', async () => {
      const channels = await getChannels();
      expect(channels).toMatchSnapshot();
    });
    it('should handle request errors', async () => {
      request.get.mockImplementation((url, cb) => cb(new Error()));
      await expect(getChannels()).rejects.toThrowError();
    });
    it('should handle non 200 responses', async () => {
      request.get.mockImplementation((url, cb) => cb(null, { statusCode: 500 }));
      await expect(getChannels()).rejects.toThrowError();
    });
  });
});
