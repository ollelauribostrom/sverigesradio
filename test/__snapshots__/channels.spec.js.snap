// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for channels.js getChannels should return a channelObject 1`] = `
Object {
  "P1": Object {
    "channeltype": "Rikskanal",
    "color": "31a1bd",
    "id": 132,
    "image": "https://static-cdn.sr.se/sida/images/132/2186745_512_512.jpg?preset=api-default-square",
    "imagetemplate": "https://static-cdn.sr.se/sida/images/132/2186745_512_512.jpg",
    "liveaudio": Object {
      "id": 132,
      "statkey": "/app/direkt/p1[k(132)]",
      "url": "https://sverigesradio.se/topsy/direkt/srapi/132.mp3",
    },
    "name": "P1",
    "scheduleurl": "https://api.sr.se/v2/scheduledepisodes?channelid=132",
    "siteurl": "https://sverigesradio.se/p1",
    "tagline": "Talat innehåll om samhälle, kultur och vetenskap. Kanalen erbjuder nyheter och aktualiteter, granskning och fördjupning men också livsåskådnings-och livsstilsprogram samt underhållning och upplevelser till exempel i form av teater.",
    "xmltvid": "p1.sr.se",
  },
  "P2": Object {
    "channeltype": "Rikskanal",
    "color": "ff5a00",
    "id": 163,
    "image": "https://static-cdn.sr.se/sida/images/163/2186754_512_512.jpg?preset=api-default-square",
    "imagetemplate": "https://static-cdn.sr.se/sida/images/163/2186754_512_512.jpg",
    "liveaudio": Object {
      "id": 163,
      "statkey": "/app/direkt/p2[k(163)]",
      "url": "https://sverigesradio.se/topsy/direkt/srapi/163.mp3",
    },
    "name": "P2",
    "scheduleurl": "https://api.sr.se/v2/scheduledepisodes?channelid=163",
    "siteurl": "https://sverigesradio.se/p2",
    "tagline": "P2 erbjuder konstmusik – klassisk såväl som nutida – improvisationsmusik och traditionsmusik, till exempel jazz eller världsmusik. P2 sänder på andra språk än svenska.",
    "xmltvid": "p2.sr.se",
  },
  "P3": Object {
    "channeltype": "Rikskanal",
    "color": "19a972",
    "id": 164,
    "image": "https://static-cdn.sr.se/sida/images/164/2186756_512_512.jpg?preset=api-default-square",
    "imagetemplate": "https://static-cdn.sr.se/sida/images/164/2186756_512_512.jpg",
    "liveaudio": Object {
      "id": 164,
      "statkey": "/app/direkt/p3[k(164)]",
      "url": "https://sverigesradio.se/topsy/direkt/srapi/164.mp3",
    },
    "name": "P3",
    "scheduleurl": "https://api.sr.se/v2/scheduledepisodes?channelid=164",
    "siteurl": "https://sverigesradio.se/p3",
    "tagline": "P3 är ung public service. Samhällsjournalistik, populärkultur, humor och relationer är bärande delar. P3 guidar till den nya musiken och lyfter fram nya svenska artister.",
    "xmltvid": "p3.sr.se",
  },
  "P4-BLEKINGE": Object {
    "channeltype": "Lokal kanal",
    "color": "c31eaa",
    "id": 213,
    "image": "https://static-cdn.sr.se/sida/images/213/2186760_512_512.jpg?preset=api-default-square",
    "imagetemplate": "https://static-cdn.sr.se/sida/images/213/2186760_512_512.jpg",
    "liveaudio": Object {
      "id": 213,
      "statkey": "/app/direkt/p4 blekinge[k(213)]",
      "url": "https://sverigesradio.se/topsy/direkt/srapi/213.mp3",
    },
    "name": "P4 Blekinge",
    "scheduleurl": "https://api.sr.se/v2/scheduledepisodes?channelid=213",
    "siteurl": "https://sverigesradio.se/blekinge/",
    "tagline": "Blekinge, P4 Blekinge, hela Blekinges lokala radiostation, Sverges Radio Blekinge",
    "xmltvid": "p4blek.sr.se",
  },
  "P4-DALARNA": Object {
    "channeltype": "Lokal kanal",
    "color": "c31eaa",
    "id": 223,
    "image": "https://static-cdn.sr.se/sida/images/223/2186762_512_512.jpg?preset=api-default-square",
    "imagetemplate": "https://static-cdn.sr.se/sida/images/223/2186762_512_512.jpg",
    "liveaudio": Object {
      "id": 223,
      "statkey": "/app/direkt/p4 dalarna[k(223)]",
      "url": "https://sverigesradio.se/topsy/direkt/srapi/223.mp3",
    },
    "name": "P4 Dalarna",
    "scheduleurl": "https://api.sr.se/v2/scheduledepisodes?channelid=223",
    "siteurl": "https://sverigesradio.se/dalarna/",
    "tagline": "mainchannel",
    "xmltvid": "p4dala.sr.se",
  },
  "P4-GOTLAND": Object {
    "channeltype": "Lokal kanal",
    "color": "c31eaa",
    "id": 205,
    "image": "https://static-cdn.sr.se/sida/images/205/2186764_512_512.jpg?preset=api-default-square",
    "imagetemplate": "https://static-cdn.sr.se/sida/images/205/2186764_512_512.jpg",
    "liveaudio": Object {
      "id": 205,
      "statkey": "/app/direkt/p4 gotland[k(205)]",
      "url": "https://sverigesradio.se/topsy/direkt/srapi/205.mp3",
    },
    "name": "P4 Gotland",
    "scheduleurl": "https://api.sr.se/v2/scheduledepisodes?channelid=205",
    "siteurl": "https://sverigesradio.se/gotland/",
    "tagline": "Mainchannel",
    "xmltvid": "p4gotl.sr.se",
  },
  "P4-GÄVLEBORG": Object {
    "channeltype": "Lokal kanal",
    "color": "c31eaa",
    "id": 210,
    "image": "https://static-cdn.sr.se/sida/images/210/2186766_512_512.jpg?preset=api-default-square",
    "imagetemplate": "https://static-cdn.sr.se/sida/images/210/2186766_512_512.jpg",
    "liveaudio": Object {
      "id": 210,
      "statkey": "/app/direkt/p4 gävleborg[k(210)]",
      "url": "https://sverigesradio.se/topsy/direkt/srapi/210.mp3",
    },
    "name": "P4 Gävleborg",
    "scheduleurl": "https://api.sr.se/v2/scheduledepisodes?channelid=210",
    "siteurl": "https://sverigesradio.se/gavleborg/",
    "tagline": "I P4 Gävleborg hör du lokala nyheter, lokala aktualiteter och lokal kultur. Du bjuds på en heltäckande, angelägen kanal där du bor.",
    "xmltvid": "p4gavl.sr.se",
  },
  "P4-GÖTEBORG": Object {
    "channeltype": "Lokal kanal",
    "color": "c31eaa",
    "id": 212,
    "image": "https://static-cdn.sr.se/sida/images/212/2186768_512_512.jpg?preset=api-default-square",
    "imagetemplate": "https://static-cdn.sr.se/sida/images/212/2186768_512_512.jpg",
    "liveaudio": Object {
      "id": 212,
      "statkey": "/app/direkt/p4 göteborg[k(212)]",
      "url": "https://sverigesradio.se/topsy/direkt/srapi/212.mp3",
    },
    "name": "P4 Göteborg",
    "scheduleurl": "https://api.sr.se/v2/scheduledepisodes?channelid=212",
    "siteurl": "https://sverigesradio.se/goteborg/",
    "tagline": "Göteborg",
    "xmltvid": "p4gbg.sr.se",
  },
  "P4-HALLAND": Object {
    "channeltype": "Lokal kanal",
    "color": "c31eaa",
    "id": 220,
    "image": "https://static-cdn.sr.se/sida/images/220/2186771_512_512.jpg?preset=api-default-square",
    "imagetemplate": "https://static-cdn.sr.se/sida/images/220/2186771_512_512.jpg",
    "liveaudio": Object {
      "id": 220,
      "statkey": "/app/direkt/p4 halland[k(220)]",
      "url": "https://sverigesradio.se/topsy/direkt/srapi/220.mp3",
    },
    "name": "P4 Halland",
    "scheduleurl": "https://api.sr.se/v2/scheduledepisodes?channelid=220",
    "siteurl": "https://sverigesradio.se/halland/",
    "tagline": "Halland",
    "xmltvid": "p4hall.sr.se",
  },
  "P4-JÄMTLAND": Object {
    "channeltype": "Lokal kanal",
    "color": "c31eaa",
    "id": 200,
    "image": "https://static-cdn.sr.se/sida/images/200/2186775_512_512.jpg?preset=api-default-square",
    "imagetemplate": "https://static-cdn.sr.se/sida/images/200/2186775_512_512.jpg",
    "liveaudio": Object {
      "id": 200,
      "statkey": "/app/direkt/p4 jämtland[k(200)]",
      "url": "https://sverigesradio.se/topsy/direkt/srapi/200.mp3",
    },
    "name": "P4 Jämtland",
    "scheduleurl": "https://api.sr.se/v2/scheduledepisodes?channelid=200",
    "siteurl": "https://sverigesradio.se/jamtland/",
    "tagline": "Sveriges Radio Jämtland, SR Jämtland",
    "xmltvid": "p4jmtl.sr.se",
  },
}
`;
